src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/ImplementIDisposableCorrectly.cs

```diff
@@ -543,75 +585,180 @@ private bool ValidateExpression(IExpressionStatementOperation expressionStatemen
+                    case OperationKind.Try:
+                        return ValidateTryOperation((ITryOperation)operation);
+                    default:
+                        return false;
                 }
+            }
 
-                if (invocationExpression.Arguments.Length != 1)
+            private bool ValidateExpression(IExpressionStatementOperation expressionStatement)
+            {
+                if (expressionStatement.Operation?.Kind != OperationKind.Invocation)
                 {
                     return false;
                 }
 
-                IOperation argumentValue = invocationExpression.Arguments[0].Value;
-                if (argumentValue.Kind != OperationKind.Conversion)
+                var invocation = (IInvocationOperation)expressionStatement.Operation;
+
+                // Valid calls are either to Dispose(false), or to the Finalize method of the base type
+                if (!_callDispose)
+                {
+                    bool result = IsDisposeBoolCall(invocation, _type, expectedValue: false);
+                    if (result)
+                    {
+                        _callDispose = true;
+                    }
+
+                    return result;
+                }
+                else if (_type.BaseType != null && invocation.Instance != null && invocation.Instance.Kind == OperationKind.InstanceReference)
+                {
+                    IMethodSymbol methodSymbol = invocation.TargetMethod;
+                    IInstanceReferenceOperation receiver = (IInstanceReferenceOperation)invocation.Instance;
+
+                    return methodSymbol.IsFinalizer() && receiver.Type.OriginalDefinition == _type.BaseType.OriginalDefinition;
+                }
+
+                return false;
+            }
+
+            private bool ValidateTryOperation(ITryOperation tryOperation)
+            {
+                // The try operation must have been implicit, as we still analyze it if it isn't implicit
+                if (!tryOperation.IsImplicit)
                 {
                     return false;
                 }
 
-                var conversion = (IConversionOperation)argumentValue;
-                if (conversion.Operand == null || conversion.Operand.Kind != OperationKind.InstanceReference)
+                if (tryOperation.Finally == null || !tryOperation.Finally.IsImplicit)
                 {
                     return false;
                 }
 
-                if (!_type.Equals(conversion.Operand.Type))
+                // Due to https://github.com/dotnet/roslyn/issues/23226, the finally block doesn't have
+                // any operations in it. When that is fixed, the code in the if will verify it.
+                if (tryOperation.Finally.Operations.Length != 0)
                 {
                     return false;
+
+#if false
+                    // The finally block must contain a single implicit call to the finalizer of the base type
```

Comment:
❓ Is it possible in C# to hit this and *not* have it be correct?

Reply:
Talked with Sam offline. We don't actually need to verify the body of this finally, as we've already determined it was compiler-generated by this point.

---
In reply to: [151559179](https://github.com/dotnet/roslyn-analyzers/pull/1371#discussion_r151559179) [](ancestors = 151559179)